File: dotnet/fsharp/tests/Domain.EventTests.fs
namespace Domain

open System
open Xunit

module EventTests =

    let Given state:InventoryItem = state
    let When event state = Apply state event
    let Then (actualState: InventoryItem) (expectedState: InventoryItem) = Assert.Equal(expectedState, actualState)

    let rdm = Random();

    [<Fact>]
    let ``Given initial inventory item, When Created event, Then Created inventory item`` () =
        let id = Guid.NewGuid()
        let initialState = id |> InventoryItem.Initialize
        let name = "Test"

        Given initialState
        |> When (Created name)
        |> Then { initialState with Name = name }

    [<Fact>]
    let ``Given inventory item, When Stocked event, Then inventory count incremented and not sold out`` () =
        let initialState = { Id = Guid.NewGuid(); Name = "Test"; Count = 0; Active = false; SoldOut = true; }
        let stockedCount = rdm.Next()

        Given initialState
        |> When (Stocked stockedCount)
        |> Then { initialState with Count = stockedCount; SoldOut = false }

    [<Fact>]
    let ``Given inventory item, When Sold event, Then inventory count decremented`` () =
        let initialState = { Id = Guid.NewGuid(); Name = "Test"; Count = 10; Active = false; SoldOut = false; }
        let soldCount = rdm.Next(10)

        Given initialState
        |> When (Sold soldCount)
        |> Then { initialState with Count = (initialState.Count - soldCount) }

    [<Fact>]
    let ``Given inventory item, When Deactivated event, Then inventory item inactive`` () =
        let initialState = { Id = Guid.NewGuid(); Name = "Test"; Count = 0; Active = true; SoldOut = true; }

        Given initialState
        |> When (Deactivated)
        |> Then { initialState with Active = false }

    [<Fact>]
    let ``Given inventory item, When Activated event, Then inventory item active`` () =
        let initialState = { Id = Guid.NewGuid(); Name = "Test"; Count = 0; Active = false; SoldOut = true; }

        Given initialState
        |> When (Activated)
        |> Then { initialState with Active = true }

    [<Fact>]
    let ``Given inventory item, When Sold more than count event, Then inventory item sold out`` () =
        let initialState = { Id = Guid.NewGuid(); Name = "Test"; Count = 9; Active = true; SoldOut = false; }

        Given initialState
        |> When (Sold 10)
        |> Then { initialState with Count = 0; SoldOut = true}

File: dotnet/csharp/tests/Domain.EventTests.cs
using Projector.Domain;
using System;
using Xunit;

namespace Domain
{
    public class EventTests
    {
        private static InventoryItem Given(InventoryItem state) => state;
        private static InventoryItem When(InventoryItem state, InventoryEvent @event) => InventoryItem.Apply(@event, state);
        private static void Then(InventoryItem expectedState, InventoryItem actualState) => Assert.Equal(expectedState, actualState);

        private static Random rdm => new Random();

        [Fact]
        public void Given_initial_inventory_item_When_Created_event_Then_Created_inventory_item()
        {
            var id = Guid.NewGuid();
            var initialState = InventoryItem.Init(id);
            string name = "Test";

            InventoryItem given = Given(initialState);
            InventoryItem when = When(given, new InventoryEvent.Created(name));
            Then(new InventoryItem(id, name, 0, false, true), when);
        }

        [Fact]
        public void Given_inventory_item_When_Stocked_event_Then_inventory_count_incremented_and_not_sold_out()
        {
            var initialState = new InventoryItem(Guid.NewGuid(), "Test", 0, false, true);
            int stockedCount = rdm.Next();

            InventoryItem given = Given(initialState);
            InventoryItem when = When(given, new InventoryEvent.Stocked(stockedCount));
            Then(new InventoryItem(initialState.Id, initialState.Name, stockedCount, false, false), when);
        }

        [Fact]
        public void Given_inventory_item_When_Sold_event_Then_inventory_count_decremented()
        {
            var initialState = new InventoryItem(Guid.NewGuid(), "Test", 10, false, false);
            int soldCount = rdm.Next(10);

            InventoryItem given = Given(initialState);
            InventoryItem when = When(given, new InventoryEvent.Sold(soldCount));
            Then(new InventoryItem(initialState.Id, initialState.Name, initialState.Count - soldCount, false, false), when);
        }

        [Fact]
        public void Given_inventory_item_When_Deactivated_event_Then_inventory_item_inactive()
        {
            var initialState = new InventoryItem(Guid.NewGuid(), "Test", 0, true, true);

            InventoryItem given = Given(initialState);
            InventoryItem when = When(given, new InventoryEvent.Deactivated());
            Then(new InventoryItem(initialState.Id, initialState.Name, initialState.Count, false, true), when);
        }

        [Fact]
        public void Given_inventory_item_When_Activated_event_Then_inventory_item_active()
        {
            var initialState = new InventoryItem(Guid.NewGuid(), "Test", 0, false, true);

            InventoryItem given = Given(initialState);
            InventoryItem when = When(given, new InventoryEvent.Activated());
            Then(new InventoryItem(initialState.Id, initialState.Name, initialState.Count, true, true), when);
        }

        [Fact]
        public void Given_inventory_item_When_Sold_more_than_count_event_Then_inventory_item_sold_out()
        {
            var initialState = new InventoryItem(Guid.NewGuid(), "Test", 9, true, false);

            InventoryItem given = Given(initialState);
            InventoryItem when = When(given, new InventoryEvent.Sold(10));
            Then(new InventoryItem(initialState.Id, initialState.Name, 0, true, true), when);
        }
    }
}